File: Sample/Engine/UI/Base/ControlBase.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
import logging
from typing import Optional, List, Union, TYPE_CHECKING
from ... import TypeAdapter, Pair, Drawable, Transformable, Vector2f, Angle, degrees, Transform, FloatRect

if TYPE_CHECKING:
    from Engine.UI import Canvas, ListView


class ControlBase(Drawable, Transformable):
    """Base class for all UI controls.

    Provides visibility, activity state, naming, parent-child relationships,
    and transform operations (position, rotation, scale, origin).
    """

    def __init__(self) -> None:
        """Construct an empty UI control with default visibility and activity."""
        Drawable.__init__(self)
        Transformable.__init__(self)
        self._visible: bool = True
        self._name: str = ""
        self._parent: Optional[Union[Canvas, ListView]] = None

    def getVisible(self) -> bool:
        r"""\brief Check whether this control is visible.

        - \return  True if visible, False otherwise
        """
        return self._visible

    def setVisible(self, visible: bool) -> None:
        r"""\brief Set the visibility of this control.

        - \param visible  True to make visible, False to hide
        """
        self._visible = visible

    def getName(self) -> str:
        r"""\brief Get the name of this control.

        - \return  Control name
        """
        return self._name

    def setName(self, name: str) -> None:
        r"""\brief Set the name of this control.

        - \param name  New control name
        """
        self._name = name

    def getParent(self) -> Optional[Union[Canvas, ListView]]:
        r"""\brief Get the parent control.

        - \return  Parent control, or None if no parent
        """
        return self._parent

    def setParent(self, parent: Optional[Union[Canvas, ListView]]) -> None:
        r"""\brief Set the parent control.

        - \param parent  New parent control, or None to detach
        """
        self._parent = parent

    def getChildren(self) -> List["ControlBase"]:
        r"""\brief Get child controls.

        - \return  Empty list for leaf controls
        """
        return []

    def getSize(self) -> Vector2f:
        r"""\brief Get the control size.

        - \return  Zero size for controls without explicit bounds
        """
        return Vector2f()

    def getLocalBounds(self) -> FloatRect:
        r"""\brief Get local bounds.

        - \return  Bounds derived from the default size
        """
        return FloatRect(Vector2f(), self.getSize())

    def getAbsoluteBounds(self) -> FloatRect:
        r"""\brief Get absolute screen-space bounds after parent transforms.

        - \return  Absolute bounds in screen space
        """
        from ... import Scale

        transform = self._getScreenRenderTransform()
        bounds = self.getLocalBounds()
        realBounds = FloatRect(bounds.position * Scale, bounds.size * Scale)
        return transform.transformRect(realBounds)

    def getRenderStates(self):
        r"""\brief Get default render states.

        - \return  Default canvas render states
        """
        from ...Utils import Render

        return Render.CanvasRenderStates()

    def v_getPosition(self) -> Pair[float]:
        r"""\brief Get the position as a plain pair.

        - \return  Position as (x, y)
        """
        result = super().getPosition()
        return (result.x, result.y)

    @TypeAdapter(position=([tuple, list], Vector2f))
    def setPosition(self, position: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Set the position of this control.

        - \param position  New position in logical UI units
        """
        super().setPosition(position)

    @TypeAdapter(offset=([tuple, list], Vector2f))
    def move(self, offset: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Move the control by an offset.

        - \param offset  Offset to add to the current position
        """
        super().move(offset)

    def v_getRotation(self) -> float:
        r"""\brief Get the rotation as a plain number (degrees).

        - \return  Rotation in degrees
        """
        result = super().getRotation()
        return result.asDegrees()

    def setRotation(self, angle: Union[Angle, float]) -> None:
        r"""\brief Set the rotation of this control.

        - \param angle  Rotation (Angle object or degrees)
        """
        if not isinstance(angle, Angle):
            angle = degrees(angle)
        super().setRotation(angle)

    def rotate(self, angle: Union[Angle, float]) -> None:
        r"""\brief Rotate the control by an offset.

        - \param angle  Rotation offset (Angle object or degrees)
        """
        if not isinstance(angle, Angle):
            angle = degrees(angle)
        super().rotate(angle)

    def v_getScale(self) -> Pair[float]:
        r"""\brief Get the scale as a plain pair.

        - \return  Scale as (x, y)
        """
        result = super().getScale()
        return (result.x, result.y)

    @TypeAdapter(scale=([tuple, list], Vector2f))
    def setScale(self, scale: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Set the scale of this control.

        - \param scale  New scale in logical UI units
        """
        super().setScale(scale)

    @TypeAdapter(factor=([tuple, list], Vector2f))
    def scale(self, factor: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Scale the control by a factor.

        - \param factor  Scale factor to multiply by
        """
        super().scale(factor)

    def v_getOrigin(self) -> Pair[float]:
        r"""\brief Get the origin as a plain pair.

        - \return  Origin as (x, y)
        """
        result = super().getOrigin()
        return (result.x, result.y)

    @TypeAdapter(origin=([tuple, list], Vector2f))
    def setOrigin(self, origin: Union[Vector2f, Pair[float], List[float]]) -> None:
        r"""\brief Set the origin of this control.

        - \param origin  New origin in logical UI units
        """
        return super().setOrigin(origin)

    def _getScreenTransform(self) -> Transform:
        transform = self.getTransform()
        if self._parent:
            return self._parent._getScreenTransform() * transform
        return transform

    def _getRenderTransform(self) -> Transform:
        return self.getTransform()

    def _getScreenRenderTransform(self) -> Transform:
        transform = self._getRenderTransform()
        if self._parent:
            return self._parent._getScreenRenderTransform() * transform
        return transform

    def __del__(self) -> None:
        r"""\brief Destructor for ControlBase.

        Logs at debug level when the object is collected.
        """
        logging.debug(f"ControlBase {self} deleted")

File: Sample/Core/Engine/src/UI/ControlBase.cpp
#include <UI/ControlBase.hpp>

#include <Runtime/EngineState.hpp>
#include <UI/FunctionalBase.hpp>
#include <Utils/Render.hpp>

#include <algorithm>
#include <array>
#include <cmath>
#include <cstdint>
#include <utility>

namespace {

std::uint8_t normalizedColourComponent(float value) {
    return static_cast<std::uint8_t>(
        std::clamp(std::lround(value * 255.0f), 0L, 255L));
}

sf::Color combinePresentationColours(
    const std::unordered_map<const void*, sf::Color>& colours) {
    std::array<float, 4> combined{1.0f, 1.0f, 1.0f, 1.0f};
    for (const auto& [source, colour] : colours) {
        static_cast<void>(source);
        combined[0] *= static_cast<float>(colour.r) / 255.0f;
        combined[1] *= static_cast<float>(colour.g) / 255.0f;
        combined[2] *= static_cast<float>(colour.b) / 255.0f;
        combined[3] *= static_cast<float>(colour.a) / 255.0f;
    }
    return {normalizedColourComponent(combined[0]),
            normalizedColourComponent(combined[1]),
            normalizedColourComponent(combined[2]),
            normalizedColourComponent(combined[3])};
}

std::uint8_t modulateColourComponent(std::uint8_t authored,
                                     std::uint8_t presentation) {
    return static_cast<std::uint8_t>(
        (static_cast<unsigned int>(authored) * presentation + 127U) / 255U);
}

}  // namespace

std::weak_ptr<RuntimeCallbackRegistry>
    ControlBase::activeRuntimeCallbackRegistry_;

void RuntimeCallbackRegistry::registerControl(ControlBase* control) {
    if (control == nullptr) {
        return;
    }
    const std::lock_guard<std::mutex> lock(mutex_);
    controls_.insert(control);
}

void RuntimeCallbackRegistry::unregisterControl(ControlBase* control) noexcept {
    const std::lock_guard<std::mutex> lock(mutex_);
    controls_.erase(control);
}

void RuntimeCallbackRegistry::releaseRuntimeCallbacks() noexcept {
    std::vector<ControlBase*> controls;
    {
        const std::lock_guard<std::mutex> lock(mutex_);
        controls.assign(controls_.begin(), controls_.end());
    }
    for (ControlBase* control : controls) {
        if (!contains(control)) {
            continue;
        }
        const std::shared_ptr<ControlBase> owner =
            control->weak_from_this().lock();
        if (owner != nullptr && owner->getParent() == nullptr) {
            owner->releaseRuntimeCallbacks();
        }
    }
}

bool RuntimeCallbackRegistry::contains(ControlBase* control) const noexcept {
    const std::lock_guard<std::mutex> lock(mutex_);
    return controls_.contains(control);
}

ControlBase::ControlBase() {
    adoptRuntimeCallbackRegistry(activeRuntimeCallbackRegistry_.lock());
}

ControlBase::~ControlBase() {
    const std::shared_ptr<RuntimeCallbackRegistry> registry =
        runtimeCallbackRegistry_.lock();
    if (registry != nullptr) {
        registry->unregisterControl(this);
    }
}

bool ControlBase::getVisible() const {
    return visible_;
}

void ControlBase::setVisible(bool visible) {
    if (visible_ == visible) {
        return;
    }
    visible_ = visible;
    if (!visible_) {
        resetFunctionalInteractions(*this);
        if (presentationRelease_) {
            presentationRelease_();
        }
    }
}

const std::string& ControlBase::getName() const {
    return name_;
}

void ControlBase::setName(const std::string& name) {
    name_ = name;
}

std::shared_ptr<ControlBase> ControlBase::getParent() const {
    return parent_.lock();
}

void ControlBase::setParent(const std::shared_ptr<ControlBase>& parent) {
    if (parent_.lock() == parent) {
        return;
    }
    resetFunctionalInteractions(*this);
    parent_ = parent;
    if (parent != nullptr) {
        const std::shared_ptr<RuntimeCallbackRegistry> registry =
            parent->runtimeCallbackRegistry_.lock();
        if (registry != nullptr) {
            adoptRuntimeCallbackRegistry(registry);
        }
    }
}

std::vector<std::shared_ptr<ControlBase>> ControlBase::getChildren() const {
    return {};
}

sf::Vector2f ControlBase::getSize() const {
    return {0.0f, 0.0f};
}

sf::FloatRect ControlBase::getLocalBounds() const {
    return {{0.0f, 0.0f}, getSize()};
}

sf::FloatRect ControlBase::getAbsoluteBounds() const {
    const sf::FloatRect bounds = getLocalBounds();
    const sf::FloatRect scaledBounds(bounds.position * Scale,
                                     bounds.size * Scale);
    return _getScreenRenderTransform().transformRect(scaledBounds);
}

sf::FloatRect ControlBase::getContentBounds() const {
    return getLocalBounds();
}

sf::FloatRect ControlBase::getAbsoluteInteractionBounds() const {
    return clipAbsoluteInteractionBounds(getAbsoluteBounds());
}

sf::FloatRect ControlBase::clipAbsoluteInteractionBounds(
    const sf::FloatRect& bounds) const {
    if (_ignoresAncestorInteractionClip()) {
        return bounds;
    }
    sf::FloatRect clipped = bounds;
    std::shared_ptr<ControlBase> parent = getParent();
    while (parent != nullptr) {
        const std::optional<sf::FloatRect> clip =
            parent->_getAbsoluteChildInteractionClipBounds();
        if (clip.has_value()) {
            const std::optional<sf::FloatRect> intersection =
                clipped.findIntersection(*clip);
            if (!intersection.has_value()) {
                return {{0.0f, 0.0f}, {0.0f, 0.0f}};
            }
            clipped = *intersection;
        }
        parent = parent->getParent();
    }
    return clipped;
}

sf::RenderStates ControlBase::getRenderStates() const {
    return canvasRenderStates();
}

void ControlBase::setPosition(const sf::Vector2f& position) {
    sf::Transformable::setPosition(position);
}

sf::Vector2f ControlBase::getPosition() const {
    return sf::Transformable::getPosition();
}

void ControlBase::move(const sf::Vector2f& offset) {
    sf::Transformable::move(offset);
}

sf::Angle ControlBase::getRotation() const {
    return sf::Transformable::getRotation();
}

void ControlBase::setRotation(sf::Angle angle) {
    sf::Transformable::setRotation(angle);
}

void ControlBase::setRotationDegrees(float angle) {
    sf::Transformable::setRotation(sf::degrees(angle));
}

void ControlBase::rotate(sf::Angle angle) {
    sf::Transformable::rotate(angle);
}

void ControlBase::rotateDegrees(float angle) {
    sf::Transformable::rotate(sf::degrees(angle));
}

sf::Vector2f ControlBase::getScale() const {
    return sf::Transformable::getScale();
}

void ControlBase::setScale(const sf::Vector2f& scale) {
    sf::Transformable::setScale(scale);
}

void ControlBase::scale(const sf::Vector2f& factor) {
    sf::Transformable::scale(factor);
}

sf::Vector2f ControlBase::getOrigin() const {
    return sf::Transformable::getOrigin();
}

void ControlBase::setOrigin(const sf::Vector2f& origin) {
    sf::Transformable::setOrigin(origin);
}

sf::Transform ControlBase::getTransform() const {
    return sf::Transformable::getTransform();
}

sf::Transform ControlBase::getInverseTransform() const {
    return sf::Transformable::getInverseTransform();
}

sf::Transform ControlBase::renderTransform() const {
    return _getRenderTransform();
}

sf::Transform ControlBase::screenRenderTransform() const {
    return _getScreenRenderTransform();
}

void ControlBase::setPresentationTransform(const sf::Vector2f& translation,
                                           float rotation,
                                           const sf::Vector2f& scale,
                                           const sf::Vector2f& pivot) {
    presentationTranslation_ = translation;
    presentationRotation_ = rotation;
    presentationScale_ = scale;
    presentationPivot_ = pivot;
}

void ControlBase::resetPresentationTransform() {
    presentationTranslation_ = {0.0f, 0.0f};
    presentationRotation_ = 0.0f;
    presentationScale_ = {1.0f, 1.0f};
    presentationPivot_ = {0.5f, 0.5f};
}

void ControlBase::setPresentationColour(const void* source,
                                        const sf::Color& colour) {
    if (source == nullptr) {
        return;
    }
    presentationColours_.insert_or_assign(source, colour);
    const sf::Color combined = combinePresentationColours(presentationColours_);
    if (presentationColour_ != combined) {
        presentationColour_ = combined;
        _refreshPresentationColour();
    }
}

void ControlBase::clearPresentationColour(const void* source) {
    if (source == nullptr || presentationColours_.erase(source) == 0) {
        return;
    }
    const sf::Color combined = combinePresentationColours(presentationColours_);
    if (presentationColour_ != combined) {
        presentationColour_ = combined;
        _refreshPresentationColour();
    }
}

void ControlBase::setPresentationUpdater(std::function<void(float)> updater) {
    presentationUpdater_ = std::move(updater);
}

void ControlBase::setPresentationRelease(std::function<void()> release) {
    presentationRelease_ = std::move(release);
}

void ControlBase::updatePresentationAnimations(float deltaTime) {
    if (presentationUpdater_) {
        presentationUpdater_(deltaTime);
    }
}

void ControlBase::refreshDisplayScale() {
    for (const std::shared_ptr<ControlBase>& child : getChildren()) {
        if (child != nullptr) {
            child->refreshDisplayScale();
        }
    }
}

void ControlBase::releaseRuntimeCallbacks() noexcept {
    if (presentationRelease_) {
        presentationRelease_();
    }
    if (FunctionalBase* functional = dynamic_cast<FunctionalBase*>(this)) {
        functional->clearEventCallbacks();
    }
    for (const std::shared_ptr<ControlBase>& child : getChildren()) {
        if (child != nullptr) {
            child->releaseRuntimeCallbacks();
        }
    }
}

void ControlBase::adoptRuntimeCallbackRegistry(
    const std::shared_ptr<RuntimeCallbackRegistry>& registry) {
    const std::shared_ptr<RuntimeCallbackRegistry> current =
        runtimeCallbackRegistry_.lock();
    if (current != registry) {
        if (current != nullptr) {
            current->unregisterControl(this);
        }
        runtimeCallbackRegistry_ = registry;
        if (registry != nullptr) {
            registry->registerControl(this);
        }
    }
    for (const std::shared_ptr<ControlBase>& child : getChildren()) {
        if (child != nullptr) {
            child->adoptRuntimeCallbackRegistry(registry);
        }
    }
}

void ControlBase::activateRuntimeCallbackRegistry(
    const std::shared_ptr<RuntimeCallbackRegistry>& registry) noexcept {
    activeRuntimeCallbackRegistry_ = registry;
}

void ControlBase::deactivateRuntimeCallbackRegistry(
    const std::shared_ptr<RuntimeCallbackRegistry>& registry) noexcept {
    if (activeRuntimeCallbackRegistry_.lock() == registry) {
        activeRuntimeCallbackRegistry_.reset();
    }
}

void ControlBase::resetActiveRuntimeCallbackRegistry() noexcept {
    activeRuntimeCallbackRegistry_.reset();
}

void ControlBase::resetFunctionalInteractions(ControlBase& control) {
    if (FunctionalBase* functional = dynamic_cast<FunctionalBase*>(&control)) {
        functional->resetPointerInteraction();
    }
    for (const std::shared_ptr<ControlBase>& child : control.getChildren()) {
        if (child != nullptr) {
            resetFunctionalInteractions(*child);
        }
    }
}

bool ControlBase::_hasOverlay() const {
    return false;
}

void ControlBase::_drawOverlay(sf::RenderTarget&, sf::RenderStates) const {}

sf::Transform ControlBase::_getScreenTransform() const {
    sf::Transform transform = getTransform();
    transform.combine(presentationTransform());
    const std::shared_ptr<ControlBase> parent = getParent();
    if (parent != nullptr) {
        transform = parent->_getScreenTransform() * transform;
    }
    return transform;
}

void ControlBase::_applyRenderStates(sf::RenderStates& states) const {
    states.transform.translate(getPosition() * (Scale - 1.0f));
    states.transform.combine(getTransform());
    states.transform.combine(presentationTransform());
}

sf::Transform ControlBase::_getRenderTransform() const {
    sf::Transform transform;
    transform.translate(getPosition() * (Scale - 1.0f));
    transform.combine(getTransform());
    transform.combine(presentationTransform());
    return transform;
}

sf::Transform ControlBase::_getScreenRenderTransform() const {
    sf::Transform transform = _getRenderTransform();
    const std::shared_ptr<ControlBase> parent = getParent();
    if (parent != nullptr) {
        transform = parent->_getScreenRenderTransform() * transform;
    }
    return transform;
}

std::optional<sf::FloatRect>
ControlBase::_getAbsoluteChildInteractionClipBounds() const {
    return std::nullopt;
}

bool ControlBase::_ignoresAncestorInteractionClip() const {
    return false;
}

void ControlBase::_refreshPresentationColour() {}

const sf::Color& ControlBase::presentationColour() const {
    return presentationColour_;
}

sf::Color ControlBase::modulatePresentationColour(
    const sf::Color& colour) const {
    return {modulateColourComponent(colour.r, presentationColour_.r),
            modulateColourComponent(colour.g, presentationColour_.g),
            modulateColourComponent(colour.b, presentationColour_.b),
            modulateColourComponent(colour.a, presentationColour_.a)};
}

sf::Transform ControlBase::presentationTransform() const {
    sf::Transform result;
    result.translate(presentationTranslation_ * Scale);
    const sf::FloatRect bounds = getLocalBounds();
    const sf::Vector2f pivot =
        (bounds.position + sf::Vector2f(bounds.size.x * presentationPivot_.x,
                                        bounds.size.y * presentationPivot_.y)) *
        Scale;
    result.translate(pivot);
    result.rotate(sf::degrees(presentationRotation_));
    result.scale(presentationScale_);
    result.translate({-pivot.x, -pivot.y});
    return result;
}

void ControlBase::draw(sf::RenderTarget&, sf::RenderStates) const {}
